Add class FEImmersedSurfaceValues

simonsticko opened the pull request:
That has the same shape function values/derivatives as FEValues,
but has JxW-values corresponding to integration over a surface
immersed in the cell. This is achieved by passing an
ImmersedSurfaceQuadrature instead of a Quadrature to the constructor.

Add a virtual function fill_fe_immersed_surface_values on Mapping to
be able to compute the correct JxW-values and surface normals. Make it
throw ExcNotImplemented by default, but override it with an
implementation on MappingCartesian.

simonsticko commented:
> One concern I have is that we have added something to `Mapping` but only implemented it for `MappingCartesian`. Is there a path forward for doing this for all the other mapping classes?

Yes, the only thing that differs is that the `JxW`-values and the normals transform as written down here:

https://www.dealii.org/current/doxygen/deal.II/classNonMatching_1_1ImmersedSurfaceQuadrature.html

which is well-defined regardless of mapping. In practice, one would almost always use a Cartesian mesh, but If you want to, I can implement the function that was added on `Mapping` on `MappingQGeneric` too. Maybe that makes sense since you get `MappingQGeneric` by default for the other `FE*Values`-classes.

peterrum commented on source/fe/mapping_cartesian.cc:
```diff
@@ -601,6 +601,90 @@ MappingCartesian<dim, spacedim>::fill_fe_subface_values(
+
+  // Convert data object to internal data for this class. Fails with an
+  // exception if that is not possible.
+  Assert(dynamic_cast<const InternalData *>(&internal_data) != nullptr,
+         ExcInternalError());
+  const InternalData &data = static_cast<const InternalData &>(internal_data);
```
```suggestion
  const auto data = dynamic_cast<const InternalData *>(&internal_data);
  Assert(data != nullptr, ExcInternalError());
```

kronbichler replied:
The original code is the same as in all other cases, and its goal is to avoid the (expensive) dynamic cast with its lookup in release mode. This might seem nit-picking but benchmarking things around 12 years ago I realized it actually pays off. :smile:

peterrum replied:
I mean why don't we create a deal.II-macro for this purpose. Dynamic casting is quite common in the library!

peterrum commented on source/fe/mapping_cartesian.cc:
```diff
@@ -601,6 +601,90 @@ MappingCartesian<dim, spacedim>::fill_fe_subface_values(
+                                      data,
+                                      output_data.quadrature_points);
+
+  if (data.update_each & update_volume_elements)
+    {
+      double det_jacobian = 1.;
+      for (unsigned int d = 0; d < dim; ++d)
+        det_jacobian *= data.cell_extents[d];
+      data.volume_element = det_jacobian;
+    }
+
+  if (data.update_each & update_jacobians)
+    for (unsigned int i = 0; i < output_data.jacobians.size(); ++i)
+      {
+        output_data.jacobians[i] = DerivativeForm<1, dim, spacedim>();
+        for (unsigned int d = 0; d < dim; ++d)
+          output_data.jacobians[i][d][d] = data.cell_extents[d];
+      }
+
+  if (data.update_each & update_inverse_jacobians)
+    for (unsigned int i = 0; i < output_data.inverse_jacobians.size(); ++i)
+      {
+        output_data.inverse_jacobians[i] = DerivativeForm<1, dim, spacedim>();
+        for (unsigned int d = 0; d < dim; ++d)
+          output_data.inverse_jacobians[i][d][d] = 1. / data.cell_extents[d];
+      }
```
These look like in the case of `fill_fe_face_values` and the two below are the new ones? Am I right?

peterrum reviewed (commented):
Overall, it looks to me good. But, I must admit I am not an expert in these classes.

kronbichler commented on include/deal.II/fe/fe_values.h:
```diff
@@ -4588,6 +4590,110 @@ class FESubfaceValues : public FEFaceValuesBase<dim, spacedim>
+ * @ingroup feaccess
+ */
+template <int dim>
+class FEImmersedSurfaceValues : public FEValuesBase<dim, dim>
```
Yes, I think it would make sense to keep things collected in that namespace for compactness. `fe_values.h` is a very large file already, so it just helps with the additional implementation. Since we do not put restrictions on the include files between different subfolders in deal.II, it is also no problem to load the additional header in the mapping classes and set the appropriate forward declaration for the `friends` specification.

kronbichler commented:
> In practice, one would almost always use a Cartesian mesh, but If you want to, I can implement the function that was added on Mapping on `MappingQGeneric` too. Maybe that makes sense since you get `MappingQGeneric` by default for the other FE*Values-classes.

I would really like to have this for `MappingQ` as well and possible also for `MappingFE` which is used by the simplices. The reason for `MappingQ` is that I imagine to compute, at some point, some overlapping meshes that are not Cartesian. Not sure if all our infrastructure supports it yet, but I think having a path forward is great.

simonsticko commented:
I think I've addressed everything here. In particular:
- Moved this class to namespace `NonMatching`
- Implemented `fill_fe_immersed_surface_values` on `MappingQ`, so one can use an cut-method even if the background mesh isn't Cartesian.

simonsticko commented on source/non_matching/fe_immersed_values.cc:
```diff
@@ -0,0 +1,184 @@
+    UpdateFlags flags = this->compute_update_flags(update_flags);
+
+    if (flags & (update_JxW_values | update_normal_vectors))
+      flags |= update_covariant_transformation;
```
Wasn't really sure if this was the right place to set this flag.

peterrum replied:
Where is `update_covariant_transformation` needed?

simonsticko replied:
When computing normals and the `JxW`-values in  `MappingQ::fill_fe_immersed_surface_values`. Line 1363 in `mapping_q.cc`

kronbichler replied:
Indeed. The covariant/contravariant transformations are the most fundamental flags to compute the Jacobians of the transformation between reference and real cell, and as such the necessary step before one can fill the JxW or normal vectors, which depend on the Jacobian.

kronbichler commented on include/deal.II/non_matching/fe_immersed_values.h:
```diff
@@ -0,0 +1,145 @@
+    reinit(
+      const TriaIterator<DoFCellAccessor<dim, dim, level_dof_access>> &cell);
+
+
```
```suggestion
```

kronbichler commented on source/non_matching/fe_immersed_values.cc:
```diff
@@ -0,0 +1,184 @@
+           (typename FEValuesBase<dim>::ExcFEDontMatch()));
+
+    this->maybe_invalidate_previous_present_cell(cell);
+    this->present_cell = {cell};
```
Does this call via the constructor https://github.com/dealii/dealii/blob/2890a38bc0909248658137c26984c6d8e9f276f8/include/deal.II/fe/fe_values.h#L3833-L3835 work for all compilers? If I remember correctly we've had trouble with some code like this for some of the compilers (but I believe the CI should reveal it if problematic).

simonsticko replied:
I don't know. This is the same way `present_cell` is reset in the other `FEValues` class, but this was recently rewritten. What does @peterrum think?

peterrum replied:
Yes. This is a plain copy from the the other `FEValues` classes. I think we use `={}` at other places in the library. I think there are even compilers that complain if you are explicitly write the type twice...

kronbichler replied:
Well, since we've had a round of testers on it since it was merged, I think we are good. I just wanted to check because I've been bitten by these before.

kronbichler commented on include/deal.II/non_matching/fe_immersed_values.h:
```diff
@@ -0,0 +1,145 @@
+                            const ImmersedSurfaceQuadrature<dim> &quadrature,
+                            const UpdateFlags                     update_flags);
+
+    /**
+     * Reinitialize the gradients, Jacobi determinants, etc for the incoming @p cell.
+     */
+    template <bool level_dof_access>
+    void
+    reinit(
+      const TriaIterator<DoFCellAccessor<dim, dim, level_dof_access>> &cell);
+
```
Do you anticipate that we never need an initialization with `typename Triangulation<dim, spacedim>::cell_iterator &cell`, as we provide in the `FEValues` class besides the `TriaIterator<DoFCellAccessor>`?

simonsticko replied:
I've added another `reinit` function and modified the test slightly to check that it works.

kronbichler reviewed (approved):
This looks great to me, thank you for your endurance. I have a few small comments and questions, nothing serious any more.

peterrum commented:
@simonsticko Could you rebase this branch. I am wondering why the undefined reference is still popping up!

peterrum commented:
> I am wondering why the undefined reference is still popping up!

Looks good now :)

peterrum commented on source/non_matching/fe_immersed_values.cc:
```diff
@@ -0,0 +1,214 @@
+  }
+
+
+#include "fe_immersed_values.inst"
```
The error looks odd. Maybe you could disable this include if doxygen is built?

simonsticko replied:
Yeah, I didn't understand if this was a doxygen bug or something else. I've added your suggestion.

peterrum replied:
Excellent, it worked! I'll have another look at the PR later today!

peterrum commented on source/non_matching/fe_immersed_values.cc:
```diff
@@ -0,0 +1,216 @@
+             " with a Mapping that is not compatible with it."));
+
+    // No FE in this cell, so no assertion necessary here.
+    this->maybe_invalidate_previous_present_cell(cell);
```
I am wondering if we should be explicit here and prevent users to reuse this class:
```suggestion
    Assert(this->present_cell==false, ExcMessage("This class should not be reused."));
```

simonsticko replied:
Yes. It's probably a mistake if you call `reinit` twice.

peterrum reviewed (approved):
I left two minor comments. I am merging; we can make adjustments in follow-up PRs.

peterrum commented:
> Implemented fill_fe_immersed_surface_values on MappingQ, so one can use an cut-method even if the background mesh isn't Cartesian.

I guess not the quadrature-point generator needs to be extended as well ;)

simonsticko commented:
> I guess not the quadrature-point generator needs to be extended as well ;)

Not if we use a discrete level set function (`DoFHandler` and a `Vector`), because the new class in #12756 transforms such a level set function to reference space in the background. So we will generate the quadrature rules in reference space, then just use these in `FEValues` or `FEImmersedSurfaceValues`.
